File: Leetcode/113.cpp
/**
 * 113. 路径总和 II  dfs
 * link：https://leetcode-cn.com/problems/path-sum-ii/
 */

class Solution
{
    vector<int> path;
    vector<vector<int>> result;
    int target;
    int curSum = 0;

public:
    vector<vector<int>> pathSum(TreeNode *root, int sum)
    {
        if (root == nullptr)
            return {};
        target = sum;
        dfs(root);
        return result;
    }

    void dfs(TreeNode *cur)
    {
        curSum += cur->val;
        path.push_back(cur->val);

        if (cur->left != nullptr)
            dfs(cur->left);
        if (cur->right != nullptr)
            dfs(cur->right);
        if (cur->left == nullptr && cur->right == nullptr)
        {
            if (curSum == target)
                result.push_back(path);
        }

        path.pop_back();
        curSum -= cur->val;
    }
};

File: Leetcode/113.py
# 113. 路径总和 II  dfs
# link：https://leetcode-cn.com/problems/path-sum-ii/


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return []
        path = []
        result = []
        curSum: int = 0

        def dfs(cur: TreeNode) -> None:
            nonlocal curSum
            curSum += cur.val
            path.append(cur.val)

            if cur.left:
                dfs(cur.left)
            if cur.right:
                dfs(cur.right)
            if not cur.left and not cur.right:
                if sum == curSum:
                    result.append(path.copy())

            curSum -= cur.val
            path.pop()

        dfs(root)
        return result
